firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/CheckForUpdateClient.java

```diff
@@ -59,56 +51,59 @@
     // forceRefresh is false to get locally cached token if available
     Task<InstallationTokenResult> installationAuthTokenTask =
         firebaseInstallationsApi.getToken(false);
 
-    Tasks.whenAllSuccess(installationIdTask, installationAuthTokenTask)
-        .addOnSuccessListener(
-            tasks -> {
-              String fid = installationIdTask.getResult();
-              InstallationTokenResult installationTokenResult =
-                  installationAuthTokenTask.getResult();
-              checkForUpdateExecutor.execute(
-                  () -> {
-                    try {
-                      AppDistributionReleaseInternal latestRelease =
-                          getLatestReleaseFromClient(
-                              fid,
-                              firebaseApp.getOptions().getApplicationId(),
-                              firebaseApp.getOptions().getApiKey(),
-                              installationTokenResult.getToken());
-                      updateOnUiThread(
-                          () -> {
-                            if (checkForUpdateTaskCompletionSource != null
-                                && !checkForUpdateTaskCompletionSource.getTask().isComplete())
-                              checkForUpdateTaskCompletionSource.setResult(latestRelease);
-                          });
-                    } catch (FirebaseAppDistributionException ex) {
-                      updateOnUiThread(() -> setCheckForUpdateTaskCompletionError(ex));
-                    }
-                  });
-            })
-        .addOnFailureListener(
-            e ->
-                setCheckForUpdateTaskCompletionError(
-                    new FirebaseAppDistributionException(
-                        Constants.ErrorMessages.AUTHENTICATION_ERROR, AUTHENTICATION_FAILURE, e)));
-
-    return checkForUpdateTaskCompletionSource.getTask();
+    this.cachedCheckForUpdate =
+        Tasks.whenAllSuccess(installationIdTask, installationAuthTokenTask)
+            .onSuccessTask(
+                checkForUpdateExecutor,
+                tasks -> {
+                  String fid = installationIdTask.getResult();
+                  InstallationTokenResult installationTokenResult =
+                      installationAuthTokenTask.getResult();
+                  try {
+                    AppDistributionReleaseInternal latestRelease =
+                        getLatestReleaseFromClient(
+                            fid,
+                            firebaseApp.getOptions().getApplicationId(),
+                            firebaseApp.getOptions().getApiKey(),
+                            installationTokenResult.getToken());
+                    return Tasks.forResult(latestRelease);
+                  } catch (FirebaseAppDistributionException ex) {
+                    return Tasks.forException(ex);
+                  }
+                })
+            .continueWithTask(
```

Comment:
I wonder if we could just call

`.continueWith( task -> task.getResult())`

According to https://developers.google.com/android/reference/com/google/android/gms/tasks/Continuation#public-methods, to "To propagate failure from the completed Task call Task.getResult() and allow the RuntimeExecutionException to propagate. The RuntimeExecutionException will be unwrapped such that the Task returned by Task.continueWith(Continuation) or Task.continueWithTask(Continuation) fails with the original exception."

Reply:
Since getLatestReleaseFromClient already fails with a NETWORK_ERROR, we might be able to just propagate it

Reply:
I tried the `task.getResult()` approach locally, and the tests failed with `java.lang.Exception: Main looper has queued unexecuted runnables.` So please disregard this suggestion